src/radical/pilot/utils/serializer.py

```diff
@@ -0,0 +1,148 @@
+
+
+def serialize_obj(obj):
+    """
+    serialize object
+    """
+    result    = None
+    exception = None
+
+    if callable(obj):
+        try:
+            result = dill.dumps(obj)
+        except Exception as e:
+            exception = e
+
+        # if we fail, then pikle it by reference
+        if result is None:
+            # see issue: https://github.com/uqfoundation/dill/issues/128
+            try:
+                result = dill.dumps(obj, byref = True)
+            except Exception as e:
+                exception = e
+
+    else:
+        try:
+            result = dill.dumps(obj, recurse = True)
+        except Exception as e:
+            exception = e
+
+        if result is None:
+            # see issue: https://github.com/uqfoundation/dill/issues/128
+            try:
+                result = dill.dumps(obj, byref = True)
+            except Exception as e:
+                exception = e
+
+    if result is None:
+        raise Exception("object %s is not serializable") from exception
+
+    return  result
+
+
+def serialize_file(obj):
+    """
+    serialize object to file
+    # FIXME: assign unique path and id for the pickled file
+    #        to avoid overwriting to the same file
+    """
+    result    = None
+    exception = None
+
+    if callable(obj):
+        try:
+            with open(_obj_file_path, 'wb') as f:
+                dill.dump(obj, f)
+                result = _obj_file_path
+
+        except Exception as e:
+            exception = e
+
+    else:
+        try:
+            with open(_obj_file_path, 'wb') as f:
+                dill.dump(obj, f, recurse = True)
+                result = _obj_file_path
+        except Exception as e:
+            exception = e
+
+    if result is None:
+        raise Exception("object is not serializable") from exception
+
+    return _obj_file_path
+
+
+
+def deserialize_file(fname):
+    """
+    Deserialize object from file
+    """
+    result = None
+
+    # check if we have a valid file
+    if not os.path.isfile(fname):
+        return None
+
+    try:
+        with open(fname, 'rb') as f:
+            result = dill.load(f)
+            if result is None:
+                raise RuntimeError('failed to deserialize')
+            return result
+
+    except Exception as e:
+        raise Exception ("failed to deserialize object from file") from e
+
+
+def deserialize_obj(obj):
+    """
+    Deserialize object from str
+    """
+    result = None
+
+    try:
+        result = dill.loads(obj)
+        if result is None:
+            raise RuntimeError('failed to deserialize')
+        return result
+
+    except Exception as e:
+        raise Exception ("failed to deserialize from object") from e
+
+
+def serialize_bson(obj):
+
+    result = codecs.encode(pickle.dumps(obj), "base64").decode()
+
+    return result
+
+
+def deserialize_bson(obj):
+
+    result = pickle.loads(codecs.decode(obj.encode(), "base64"))
+
+    return result
```

Comment:
one more thing (sorry for many comments...) - why `pickle` is used here? and not `dill`? (kinda, it is good to have one approach, and as I understand `dill` is the same as `pickle`, but provides some extra functionality, so we can stick just to one of them)